File: sdks/android/mobile-api-studio/src/main/kotlin/dev/mobileapistudio/sdk/MobileAPIStudio.kt
package dev.mobileapistudio.sdk

import android.content.Context
import okhttp3.Request
import okhttp3.Response
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

public object MobileAPIStudio {
    public const val sdkVersion: String = "0.1.0"
    public const val correlationHeader: String = "X-Mobile-API-Studio-Request-Id"

    private data class InFlight(
        val startedAtMillis: Long,
        val method: String,
        val url: String,
        val context: MobileAPIStudioContext,
    )

    private val lock = Any()
    private var configuration: ResolvedConfiguration? = null
    private var transport: MobileAPIStudioTransport? = null
    private var clientId: String? = null
    private var currentContext = MobileAPIStudioContext()
    private val inFlight = ConcurrentHashMap<String, InFlight>()

    public val isEnabled: Boolean
        get() = synchronized(lock) { configuration?.enabled == true }

    public fun configure(
        context: Context,
        configuration: MobileAPIStudioConfiguration = MobileAPIStudioConfiguration(),
    ) {
        val appContext = context.applicationContext
        val resolved = configuration.resolved(appContext)
        val previousTransport = synchronized(lock) {
            val previous = transport
            this.configuration = resolved
            if (!resolved.enabled) {
                transport = null
                clientId = null
                currentContext = MobileAPIStudioContext()
                inFlight.clear()
                previous
            } else {
                val nextClientId = loadClientId(appContext, resolved.appId)
                transport = MobileAPIStudioTransport(resolved.desktopHost, resolved.desktopPort)
                clientId = nextClientId
                previous
            }
        }
        previousTransport?.close()
        if (resolved.enabled) sendHandshake()
    }

    public fun disable() {
        val previous = synchronized(lock) {
            configuration = configuration?.copy(enabled = false)
            val value = transport
            transport = null
            clientId = null
            currentContext = MobileAPIStudioContext()
            inFlight.clear()
            value
        }
        previous?.close()
    }

    public fun setContext(
        screen: String? = null,
        feature: String? = null,
        attributes: Map<String, String> = emptyMap(),
        source: MobileAPIStudioSource? = callerSource(),
    ) {
        val value = MobileAPIStudioContext(screen, feature, attributes, source)
        val payload = synchronized(lock) {
            currentContext = value
            val id = clientId ?: return@synchronized null
            if (configuration?.enabled != true) return@synchronized null
            mapOf("clientId" to id, "context" to value.toWire())
        }
        payload?.let { send("context", it) }
    }

    public fun log(
        message: String,
        level: MobileAPIStudioLogLevel = MobileAPIStudioLogLevel.INFO,
        attributes: Map<String, String> = emptyMap(),
        source: MobileAPIStudioSource? = callerSource(),
    ) {
        val payload = synchronized(lock) {
            val id = clientId ?: return@synchronized null
            if (configuration?.enabled != true) return@synchronized null
            val context = mergedContext(feature = null, attributes = attributes, source = source)
            mapOf(
                "clientId" to id,
                "level" to level.wireValue,
                "message" to message,
                "context" to context.toWire(),
            )
        }
        payload?.let { send("log", it) }
    }

    public fun instrument(
        request: Request,
        feature: String? = null,
        attributes: Map<String, String> = emptyMap(),
        source: MobileAPIStudioSource? = callerSource(),
    ): MobileAPIStudioInstrumentedRequest? {
        val requestId = beginRequest(
            method = request.method,
            url = request.url.toString(),
            feature = feature,
            attributes = attributes,
            source = source,
        ) ?: return null
        val instrumented = request.newBuilder()
            .header(correlationHeader, requestId)
            .build()
        return MobileAPIStudioInstrumentedRequest(requestId, instrumented)
    }

    /**
     * Starts correlation for a custom networking stack. Add the returned value to
     * [correlationHeader] on the real outgoing request, then call [complete].
     */
    public fun beginRequest(
        method: String,
        url: String,
        feature: String? = null,
        attributes: Map<String, String> = emptyMap(),
        source: MobileAPIStudioSource? = callerSource(),
    ): String? {
        val result = synchronized(lock) {
            val id = clientId ?: return@synchronized null
            if (configuration?.enabled != true) return@synchronized null
            val requestId = UUID.randomUUID().toString().lowercase()
            val context = mergedContext(feature, attributes, source)
            val normalizedMethod = method.ifBlank { "GET" }.uppercase()
            inFlight[requestId] = InFlight(
                startedAtMillis = System.currentTimeMillis(),
                method = normalizedMethod,
                url = url,
                context = context,
            )
            requestId to mapOf<String, Any?>(
                "clientId" to id,
                "requestId" to requestId,
                "phase" to "started",
                "method" to normalizedMethod,
                "url" to url,
                "context" to context.toWire(),
            )
        } ?: return null
        send("network", result.second)
        return result.first
    }

    public fun complete(
        requestId: String,
        response: Response? = null,
        error: Throwable? = null,
    ) {
        complete(
            requestId = requestId,
            statusCode = response?.code,
            error = error,
        )
    }

    public fun complete(
        requestId: String,
        statusCode: Int? = null,
        error: Throwable? = null,
    ) {
        val flight = inFlight.remove(requestId) ?: return
        val payload = synchronized(lock) {
            val id = clientId ?: return@synchronized null
            if (configuration?.enabled != true) return@synchronized null
            buildMap<String, Any?> {
                put("clientId", id)
                put("requestId", requestId)
                put("phase", if (error == null) "completed" else "failed")
                put("method", flight.method)
                put("url", flight.url)
                put("durationMs", (System.currentTimeMillis() - flight.startedAtMillis).coerceAtLeast(0))
                put("context", flight.context.toWire())
                statusCode?.let { put("statusCode", it) }
                error?.let { put("error", it.toString()) }
            }
        }
        payload?.let { send("network", it) }
    }

    public fun sourceFromStack(): MobileAPIStudioSource? = callerSource()

    private fun sendHandshake() {
        val payload = synchronized(lock) {
            val config = configuration ?: return@synchronized null
            val id = clientId ?: return@synchronized null
            if (!config.enabled) return@synchronized null
            buildMap<String, Any?> {
                put("clientId", id)
                put("appId", config.appId)
                put("appName", config.appName)
                put("platform", "android")
                put("sdkVersion", sdkVersion)
                config.appVersion?.let { put("appVersion", it) }
                config.appBuild?.let { put("appBuild", it) }
                config.deviceName?.let { put("deviceName", it) }
                config.osVersion?.let { put("osVersion", it) }
            }
        }
        payload?.let { send("handshake", it) }
    }

    private fun send(type: String, payload: Map<String, Any?>) {
        synchronized(lock) { transport }?.send(type, payload)
    }

    private fun mergedContext(
        feature: String?,
        attributes: Map<String, String>,
        source: MobileAPIStudioSource?,
    ): MobileAPIStudioContext {
        return currentContext.copy(
            feature = feature ?: currentContext.feature,
            attributes = currentContext.attributes + attributes,
            source = source ?: currentContext.source,
        )
    }

    private fun loadClientId(context: Context, appId: String): String {
        val preferences = context.getSharedPreferences("mobile-api-studio", Context.MODE_PRIVATE)
        val key = "client-id.$appId"
        val existing = preferences.getString(key, null)
        if (!existing.isNullOrBlank()) return existing
        val generated = UUID.randomUUID().toString().lowercase()
        preferences.edit().putString(key, generated).apply()
        return generated
    }

    private fun callerSource(): MobileAPIStudioSource? {
        val frame = Throwable().stackTrace.firstOrNull { element ->
            !element.className.startsWith("dev.mobileapistudio.sdk.") &&
                !element.className.startsWith("java.") &&
                !element.className.startsWith("kotlin.")
        } ?: return null
        return MobileAPIStudioSource(
            file = frame.fileName,
            function = frame.methodName,
            line = frame.lineNumber.takeIf { it >= 0 },
        )
    }
}

File: sdks/ios/MobileAPIStudio/Sources/MobileAPIStudio/MobileAPIStudio.swift
import Foundation
#if os(iOS)
import UIKit
#endif

public enum MobileAPIStudio {
    public static let sdkVersion = "0.1.0"
    public static let correlationHeader = "X-Mobile-API-Studio-Request-Id"

    public static var defaultEnabled: Bool {
        #if DEBUG
        true
        #else
        false
        #endif
    }

    private static let state = MobileAPIStudioState()

    public static func configure(_ configuration: MobileAPIStudioConfiguration = .init()) {
        state.configure(configuration)
    }

    public static func setContext(
        screen: String? = nil,
        feature: String? = nil,
        attributes: [String: String] = [:],
        file: String = #fileID,
        function: String = #function,
        line: UInt = #line
    ) {
        state.setContext(
            MobileAPIStudioContext(
                screen: screen,
                feature: feature,
                attributes: attributes,
                source: MobileAPIStudioSource(file: file, function: function, line: line)
            )
        )
    }

    public static func log(
        _ message: String,
        level: MobileAPIStudioLogLevel = .info,
        attributes: [String: String] = [:],
        file: String = #fileID,
        function: String = #function,
        line: UInt = #line
    ) {
        state.log(
            message,
            level: level,
            attributes: attributes,
            source: MobileAPIStudioSource(file: file, function: function, line: line)
        )
    }

    @discardableResult
    public static func instrument(
        _ request: inout URLRequest,
        feature: String? = nil,
        attributes: [String: String] = [:],
        file: String = #fileID,
        function: String = #function,
        line: UInt = #line
    ) -> String? {
        state.instrument(
            &request,
            feature: feature,
            attributes: attributes,
            source: MobileAPIStudioSource(file: file, function: function, line: line)
        )
    }

    public static func instrumented(
        _ request: URLRequest,
        feature: String? = nil,
        attributes: [String: String] = [:],
        file: String = #fileID,
        function: String = #function,
        line: UInt = #line
    ) -> MobileAPIStudioInstrumentedRequest? {
        var copy = request
        guard let requestID = instrument(
            &copy,
            feature: feature,
            attributes: attributes,
            file: file,
            function: function,
            line: line
        ) else {
            return nil
        }
        return MobileAPIStudioInstrumentedRequest(requestID: requestID, request: copy)
    }

    public static func complete(
        requestID: String,
        response: URLResponse? = nil,
        error: Error? = nil
    ) {
        state.complete(requestID: requestID, response: response, error: error)
    }

    public static func instrument(_ configuration: URLSessionConfiguration) -> URLSessionConfiguration {
        guard state.isEnabled else { return configuration }
        var classes = configuration.protocolClasses ?? []
        if !classes.contains(where: { $0 == MobileAPIStudioURLProtocol.self }) {
            classes.insert(MobileAPIStudioURLProtocol.self, at: 0)
            configuration.protocolClasses = classes
        }
        return configuration
    }

    internal static func autoInstrument(_ request: URLRequest) -> MobileAPIStudioInstrumentedRequest? {
        var copy = request
        guard let requestID = state.instrument(
            &copy,
            feature: nil,
            attributes: [:],
            source: nil
        ) else { return nil }
        return MobileAPIStudioInstrumentedRequest(requestID: requestID, request: copy)
    }
}

private final class MobileAPIStudioState: @unchecked Sendable {
    private struct InFlight {
        var startedAt: Date
        var method: String
        var url: String
        var context: MobileAPIStudioContext
    }

    private let lock = NSLock()
    private var configuration: MobileAPIStudioConfiguration?
    private var transport: MobileAPIStudioTransport?
    private var currentContext = MobileAPIStudioContext()
    private var inFlight: [String: InFlight] = [:]
    private var clientID: String?

    var isEnabled: Bool {
        lock.withLock { configuration?.enabled == true }
    }

    func configure(_ configuration: MobileAPIStudioConfiguration) {
        lock.withLock {
            self.configuration = configuration
            guard configuration.enabled else {
                transport = nil
                inFlight.removeAll()
                return
            }
            transport = MobileAPIStudioTransport(baseURL: configuration.desktopBaseURL)
            clientID = loadClientID(appID: configuration.appID)
        }
        sendHandshake()
    }

    func setContext(_ context: MobileAPIStudioContext) {
        let payload: [String: Any]? = lock.withLock {
            currentContext = context
            guard let clientID, configuration?.enabled == true else { return nil }
            return ["clientId": clientID, "context": context.wireValue]
        }
        if let payload { send(type: "context", payload: payload) }
    }

    func log(
        _ message: String,
        level: MobileAPIStudioLogLevel,
        attributes: [String: String],
        source: MobileAPIStudioSource
    ) {
        let payload: [String: Any]? = lock.withLock {
            guard let clientID, configuration?.enabled == true else { return nil }
            var context = currentContext
            context.attributes.merge(attributes) { _, new in new }
            context.source = source
            return [
                "clientId": clientID,
                "level": level.rawValue,
                "message": message,
                "context": context.wireValue
            ]
        }
        if let payload { send(type: "log", payload: payload) }
    }

    func instrument(
        _ request: inout URLRequest,
        feature: String?,
        attributes: [String: String],
        source: MobileAPIStudioSource?
    ) -> String? {
        let result: (String, [String: Any])? = lock.withLock {
            guard let clientID, configuration?.enabled == true,
                  let url = request.url else { return nil }
            let requestID = UUID().uuidString.lowercased()
            request.setValue(requestID, forHTTPHeaderField: MobileAPIStudio.correlationHeader)
            var context = currentContext
            if let feature { context.feature = feature }
            context.attributes.merge(attributes) { _, new in new }
            if let source { context.source = source }
            let method = request.httpMethod ?? "GET"
            inFlight[requestID] = InFlight(
                startedAt: Date(),
                method: method,
                url: url.absoluteString,
                context: context
            )
            return (
                requestID,
                [
                    "clientId": clientID,
                    "requestId": requestID,
                    "phase": "started",
                    "method": method,
                    "url": url.absoluteString,
                    "context": context.wireValue
                ]
            )
        }
        guard let result else { return nil }
        send(type: "network", payload: result.1)
        return result.0
    }

    func complete(requestID: String, response: URLResponse?, error: Error?) {
        let payload: [String: Any]? = lock.withLock {
            guard let clientID,
                  configuration?.enabled == true,
                  let flight = inFlight.removeValue(forKey: requestID) else { return nil }
            var value: [String: Any] = [
                "clientId": clientID,
                "requestId": requestID,
                "phase": error == nil ? "completed" : "failed",
                "method": flight.method,
                "url": flight.url,
                "durationMs": Int(Date().timeIntervalSince(flight.startedAt) * 1000),
                "context": flight.context.wireValue
            ]
            if let response = response as? HTTPURLResponse {
                value["statusCode"] = response.statusCode
            }
            if let error { value["error"] = String(describing: error) }
            return value
        }
        if let payload { send(type: "network", payload: payload) }
    }

    private func sendHandshake() {
        let payload: [String: Any]? = lock.withLock {
            guard let configuration,
                  configuration.enabled,
                  let clientID else { return nil }
            var value: [String: Any] = [
                "clientId": clientID,
                "appId": configuration.appID,
                "appName": configuration.appName,
                "platform": "ios",
                "sdkVersion": MobileAPIStudio.sdkVersion
            ]
            if let appVersion = configuration.appVersion { value["appVersion"] = appVersion }
            if let appBuild = configuration.appBuild { value["appBuild"] = appBuild }
            #if os(iOS)
            value["deviceName"] = UIDevice.current.name
            value["osVersion"] = UIDevice.current.systemVersion
            #else
            value["deviceName"] = Host.current().localizedName
            value["osVersion"] = ProcessInfo.processInfo.operatingSystemVersionString
            #endif
            return value
        }
        if let payload { send(type: "handshake", payload: payload) }
    }

    private func send(type: String, payload: [String: Any]) {
        let currentTransport = lock.withLock { transport }
        currentTransport?.send(type: type, payload: payload)
    }

    private func loadClientID(appID: String) -> String {
        let key = "dev.mobileapistudio.client-id.\(appID)"
        if let existing = UserDefaults.standard.string(forKey: key), !existing.isEmpty {
            return existing
        }
        let generated = UUID().uuidString.lowercased()
        UserDefaults.standard.set(generated, forKey: key)
        return generated
    }
}

private extension NSLock {
    func withLock<T>(_ body: () -> T) -> T {
        lock()
        defer { unlock() }
        return body()
    }
}
